File: app/src/main/java/com/attendx/app/data/local/entity/AttendanceRecord.kt
package com.attendx.app.data.local.entity

import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "attendance_records",
    foreignKeys = [
        ForeignKey(
            entity = Subject::class,
            parentColumns = ["id"],
            childColumns = ["scheduledSubjectId"],
            onDelete = ForeignKey.SET_NULL
        ),
        ForeignKey(
            entity = Subject::class,
            parentColumns = ["id"],
            childColumns = ["actualSubjectId"],
            onDelete = ForeignKey.SET_NULL
        )
    ],
    indices = [
        Index("scheduledSubjectId"),
        Index("actualSubjectId"),
        Index("date"),
        Index("date", "periodNumber", unique = true)
    ]
)
data class AttendanceRecord(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val date: Long, // epoch millis (date only, time set to 00:00)
    val dayOfWeek: Int,
    val periodNumber: Int,
    val scheduledSubjectId: Long? = null,
    val actualSubjectId: Long? = null,
    val status: String, // PRESENT, ABSENT, CANCELLED, FREE, SEMINAR
    val note: String? = null,
    val createdAt: Long = System.currentTimeMillis()
) {
    companion object {
        const val STATUS_PRESENT = "PRESENT"
        const val STATUS_ABSENT = "ABSENT"
        const val STATUS_CANCELLED = "CANCELLED"
        const val STATUS_FREE = "FREE"
        const val STATUS_SEMINAR = "SEMINAR"
        const val STATUS_HOLIDAY = "HOLIDAY"
    }
}

File: lib/models/attendance_record.dart
class AttendanceRecord {
  final int? id;
  final int date; // Epoch millis, stored as day start
  final int dayOfWeek;
  final int periodNumber;
  final int? scheduledSubjectId;
  final int? actualSubjectId;
  final String status; // 'PRESENT', 'ABSENT', 'CANCELLED'
  final String? note;
  final int createdAt;

  AttendanceRecord({
    this.id,
    required this.date,
    required this.dayOfWeek,
    required this.periodNumber,
    this.scheduledSubjectId,
    this.actualSubjectId,
    required this.status,
    this.note,
    required this.createdAt,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'date': date,
      'dayOfWeek': dayOfWeek,
      'periodNumber': periodNumber,
      'scheduledSubjectId': scheduledSubjectId,
      'actualSubjectId': actualSubjectId,
      'status': status,
      'note': note,
      'createdAt': createdAt,
    };
  }

  factory AttendanceRecord.fromMap(Map<String, dynamic> map) {
    return AttendanceRecord(
      id: map['id'] as int?,
      date: map['date'] as int,
      dayOfWeek: map['dayOfWeek'] as int,
      periodNumber: map['periodNumber'] as int,
      scheduledSubjectId: map['scheduledSubjectId'] as int?,
      actualSubjectId: map['actualSubjectId'] as int?,
      status: map['status'] as String,
      note: map['note'] as String?,
      createdAt: map['createdAt'] as int,
    );
  }
}
